### PR TITLE
Add bridge on icon

### PR DESCRIPTION
**Pull request by prin-r**
## Implementation details
Implement a bridge smart contract in ICON Network that mimics the bridge that is already in EVM

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [ ] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [x] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [ ] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: CHANGELOG_UNRELEASED.md [hunk 1]
```diff
@@ -21,6 +21,7 @@
 ### Scan
 
 ### Bridges
+- (feat) [\#2385](https://github.com/bandprotocol/bandchain/pull/2385) Add icon bridge.
```
Comment (taobun) [inline CHANGELOG_UNRELEASED.md:41]: Add a new line between topic and change(Other editors will format make diff)

File: bridges/icon/README.md [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+```
+tbears test bridge
+```
+
+## Deployment
+
+#### Bridge
+
+```
+tbears deploy bridge -k mumustorekey.key -c bridge/tbears_cli_config_testnet.json
```
Comment (taobun) [inline bridges/icon/README.md]: wrong file json?

File: bridges/icon/bridge/bridge.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+        # a single integer
+        self.total_validator_power = VarDB("total_validator_power", db, value_type=int)
+
+        # pubkey:bytes => voting_power:int
+        self.validator_powers = DictDB("validator_powers", db, value_type=int)
+
+        # block_number:int => oracle_state_hash:bytes
+        self.oracle_state = DictDB("oracle_state", db, value_type=bytes)
+
+        # encoded_request:bytes => response:dict
```
Comment (taobun) [inline bridges/icon/bridge/bridge.py]: It seems to be bytes of obi encoded response packet.

**Review by taobun — CHANGES_REQUESTED** [review]

File: bridges/icon/bridge/bridge.py (hunk 3 above)
Comment (prin-r) [reply bridges/icon/bridge/bridge.py]: done

File: bridges/icon/README.md (hunk 2 above)
Comment (prin-r) [reply bridges/icon/README.md]: done

File: CHANGELOG_UNRELEASED.md (hunk 1 above)
Comment (prin-r) [reply CHANGELOG_UNRELEASED.md:41]: done

File: bridges/icon/bridge/bridge.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        # a single integer
+        self.total_validator_power = VarDB("total_validator_power", db, value_type=int)
+
+        # pubkey:bytes => voting_power:int
+        self.validator_powers = DictDB("validator_powers", db, value_type=int)
+
+        # block_number:int => oracle_state_hash:bytes
+        self.oracle_state = DictDB("oracle_state", db, value_type=bytes)
+
+	# encoded_request:bytes => encoded_response:bytes
```
Comment (taobun) [inline bridges/icon/bridge/bridge.py]: Indent

File: bridges/icon/bridge/utils/utils.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+def merkle_inner_hash(left: bytes, right: bytes) -> bytes:
+    return sha256.digest(bytes([1]) + left + right)
+
+
+def encode_varint_unsigned(value: int) -> bytes:
+    temp_value = value
+    size = 0
+    while temp_value > 0:
+        size += 1
+        temp_value >>= 7
```
Comment (taobun) [inline bridges/icon/bridge/utils/utils.py]: I think we don't need to calculate size of data first in Python.

File: bridges/icon/bridge/utils/merkle_part.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+#  [4] - last_block_id         [5] - last_commit_hash    [6] - data_hash     [7] - validators_hash
+#  [8] - next_validators_hash  [9] - consensus_hash      [A] - app_hash      [B] - last_results_hash
+#  [C] - evidence_hash         [D] - proposer_address
+#
+# Notice that NOT all leaves of the Merkle tree are needed in order to compute the Merkle
+# root hash, since we only want to validate the correctness of [A] and [2]. In fact, only
+# [1A], [3], [2B], [1E], [B], and [2D] are needed in order to compute [BlockHeader].
+
+
+def get_block_header(data: bytes, app_hash: bytes, block_height: int):
```
Comment (taobun) [inline bridges/icon/bridge/utils/merkle_part.py]: Why this function doesn't declare output type?

File: bridges/icon/bridge/utils/merkle_part.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                    data[32:64],  # [3]
+                ),
+            ),
+            data[64:96],  # [2B]
+        ),
+        utils.merkle_inner_hash(  # [3B]
+            utils.merkle_inner_hash(  # [2C]
+                data[96:128],  # [1E]
+                utils.merkle_inner_hash(  # [1F]
+                    utils.merkle_leaf_hash(bytes([32]) + app_hash), data[128:160]  # [A]  # [B]
```
Comment (taobun) [inline bridges/icon/bridge/utils/merkle_part.py]: `# [A], [B]`

File: bridges/icon/bridge/utils/multi_store.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+                utils.merkle_inner_hash(  # [I6]
+                    main_and_mint_stores_merkle_hash,  # [I3]
+                    utils.merkle_inner_hash(
+                        utils.merkle_leaf_hash(  # [I4]
+                            # [6]
+                            # oracle prefix (uint8(6) + "oracle" + uint8(32))
+                            bytes.fromhex("066f7261636c6520") +
+                            sha256.digest(sha256.digest(
+                                oracle_iavl_state_hash))
+                        ),
```
Comment (taobun) [inline bridges/icon/bridge/utils/multi_store.py]: I think to move `# [6]` at this line make comment more consistent.

**Review by taobun — CHANGES_REQUESTED** [review]

File: bridges/icon/bridge/utils/multi_store.py (hunk 8 above)
Comment (prin-r) [reply bridges/icon/bridge/utils/multi_store.py]: done

File: bridges/icon/bridge/bridge.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
+from iconservice import *
+from .utils import *
+from .pyobi import *
+
+TAG = "BRIDGE"
+
+
+class BRIDGE(IconScoreBase):
```
Comment (taobun) [inline bridges/icon/bridge/bridge.py]: Bridge

File: bridges/icon/bridge/bridge.py [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+                }
+            }
+            """
+        ).decode(encode_packet)
+
+        current_merkle_hash = sha256.digest(
+            # Height of tree (only leaf node) is 0 (signed-varint encode)
+            bytes([0])
+            + bytes([2])
+            + utils.encode_varint_signed(version)  # Size of subtree is 1 (signed-varint encode)
```
Comment (taobun) [inline bridges/icon/bridge/bridge.py]: Bring a comment to previous line.

File: bridges/icon/bridge/bridge.py [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+            +
+            # Size of data key (1-byte constant 0x01 + 8-byte request ID)
+            bytes([9])
+            + b"\xff"
+            + packet["res"][  # Constant 0xff prefix data request info storage key
```
Comment (taobun) [inline bridges/icon/bridge/bridge.py]: ditto

File: bridges/icon/bridge/bridge.py [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+                "request_id"
+            ].to_bytes(8, "big")
+            + bytes([32])
+            + sha256.digest(encode_packet)  # Size of data hash
```
Comment (taobun) [inline bridges/icon/bridge/bridge.py]: ditto

File: bridges/icon/bridge/bridge.py [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+                prev_res.get("resolve_time", 0)
+                if isinstance(prev_res.get("resolve_time", 0), int)
+                else 0
+            )
+            if isinstance(prev_res, dict)
+            else 0
+        )
+
+        if prev_resolve_time >= res["resolve_time"]:
+            self.revert("FAIL_LATEST_REQUEST_SHOULD_BE_NEWEST")
```
Comment (taobun) [inline bridges/icon/bridge/bridge.py]: REQUEST_HAS_OUTDATED

**Review by taobun — CHANGES_REQUESTED** [review]

File: bridges/icon/bridge/bridge.py (hunk 9 above)
Comment (prin-r) [reply bridges/icon/bridge/bridge.py]: done

File: bridges/icon/cache_consumer_mock/cache_consumer_mock.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+TAG = "CACHE_CONSUMER_MOCK"
+
+
+class IBridgeCache(InterfaceScore):
+    @interface
+    def get_latest_response(self, encoded_request: bytes) -> dict:
+        pass
+
+
+class CACHE_CONSUMER_MOCK(IconScoreBase):
```
Comment (taobun) [inline bridges/icon/cache_consumer_mock/cache_consumer_mock.py]: I think we can use Pascal case

File: bridges/icon/cache_consumer_mock/tests/test_unit_cache_consumer_mock.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
+from ..cache_consumer_mock import CACHE_CONSUMER_MOCK
+from tbears.libs.scoretest.score_test_case import ScoreTestCase
+
+
+class TestCACHE_CONSUMER_MOCK(ScoreTestCase):
```
Comment (taobun) [inline bridges/icon/cache_consumer_mock/tests/test_unit_cache_consumer_mock.py]: ditto

File: bridges/icon/receiver_mock/tests/test_unit_receiver_mock.py [hunk 16]
```diff
@@ -0,0 +1,9 @@
+from ..receiver_mock import RECEIVER_MOCK
+from tbears.libs.scoretest.score_test_case import ScoreTestCase
+
+
+class TestRECEIVER_MOCK(ScoreTestCase):
```
Comment (taobun) [inline bridges/icon/receiver_mock/tests/test_unit_receiver_mock.py]: ditto

File: bridges/icon/receiver_mock/tests/test_unit_receiver_mock.py [hunk 17]
```diff
@@ -0,0 +1,9 @@
...
+    def setUp(self):
+        super().setUp()
+        self.score = self.get_score_instance(RECEIVER_MOCK, self.test_account1)
```
Comment (taobun) [inline bridges/icon/receiver_mock/tests/test_unit_receiver_mock.py]: here too.

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**